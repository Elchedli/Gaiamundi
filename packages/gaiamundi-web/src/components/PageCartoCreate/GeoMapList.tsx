import { useQuery } from 'react-query';
import { useState } from 'react';
import { ApiErrorAlert } from 'components/Alert/ApiErrorMessage';
import { LoadingMessage } from 'components/Loader/LoadingMessage';
import { ContentType, strapi } from 'services/strapi';
import { Alert } from 'components/Alert/Alert';
import { ApiError } from 'interfaces/api';
import { Pagination } from 'components/Pagination/Pagination';
import GeoListItem from './GeoListItem';

export const GeoMapList = () => {
  const [page, setPage] = useState(1);
  const paginationLimit = 9;
  const {
    data: response,
    isError,
    error,
    isLoading,
  } = useQuery({
    queryKey: ['latest-geo-carto', page],
    queryFn: () => {
      return strapi.get(ContentType.GEO_MAPS, {
        populate: '*',
        sort: 'createdAt:desc',
        pagination: {
          page: page,
          pageSize: paginationLimit,
        },
      });
    },
    keepPreviousData: true,
  });

  if (isLoading) {
    return <LoadingMessage />;
  }

  if (isError) {
    return <ApiErrorAlert error={error as ApiError} />;
  }

  if (response?.data.length === 0) {
    return <Alert type="info">Aucun contenu Ã  afficher.</Alert>;
  }

  const totalPages = response?.meta.pagination.total;
  return (
    <div>
      <div className="grid grid-cols-3 gap-y-10 gap-x-6">
        {response?.data.map((page) => {
          return <GeoListItem key={page.id} {...page} />;
        })}
      </div>
      <div className="flex flex-row mt-5 justify-center">
        <Pagination
          page={page}
          onPaginateNext={() => setPage(page + 1)}
          onPaginatePrevious={() => setPage(page - 1)}
          onPaginate={(p: number) => setPage(p)}
          totalPages={Math.ceil(totalPages! / paginationLimit)}
        />
      </div>
    </div>
  );
};
