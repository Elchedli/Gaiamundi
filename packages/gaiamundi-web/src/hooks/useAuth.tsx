import React from 'react';
import {
  QueryObserverResult,
  RefetchOptions,
  UseMutateFunction,
  useQuery,
  useQueryClient,
} from 'react-query';

import { Alert } from 'components/Alert/Alert';
import { LoadingMessage } from 'components/Loader/LoadingMessage';
import { User } from 'interfaces/user';
import { strapi } from '../services/strapi';
import { useLocalStorageState } from './useLocalStorageState';

// const updateAccount = async (id: number, userData: UserSignUpFields) => {
//   return (await strapi.update(ContentType.USERS, id, userData)) as User;
// };

export interface AuthContextValue {
  user: User | undefined;
  isAuthenticated: boolean;
  setUser: (data: User | undefined) => void;
  authenticate: (user: User, jwt: string) => void;
  logout: UseMutateFunction;
  refetchUser: (
    options?: RefetchOptions | undefined
  ) => Promise<QueryObserverResult<User, Error>>;
  error: Error | null;
}

const AuthContext = React.createContext<AuthContextValue | null>(null);
AuthContext.displayName = 'AuthContext';

export interface AuthProviderProps {
  children: React.ReactNode;
}

export const AuthProvider = ({ children }: AuthProviderProps): JSX.Element => {
  const { value: jwtToken, persistValue: persistJwtToken } =
    useLocalStorageState('token');
  const queryClient = useQueryClient();
  const key = 'auth-user';

  const {
    data: user,
    error,
    isLoading,
    refetch,
  } = useQuery<User, Error>({
    queryKey: key,
    queryFn: async () => {
      return await strapi.currentUser(jwtToken);
    },
    enabled: !!jwtToken,
  });

  const setUser = (data: User | undefined) =>
    queryClient.setQueryData(key, data);

  const authenticate = (user: User, jwt: string) => {
    setUser(user);
    persistJwtToken(jwt);
  };

  const logout = () => {
    persistJwtToken('');
    queryClient.clear();
  };

  if (isLoading) {
    return <LoadingMessage />;
  }

  return (
    <AuthContext.Provider
      value={{
        user,
        isAuthenticated: !!jwtToken,
        error,
        setUser,
        authenticate,
        refetchUser: refetch,
        logout,
      }}
    >
      {error && <Alert>{error.message}</Alert>}
      {children}
    </AuthContext.Provider>
  );
};

export const useAuth = () => {
  const context = React.useContext(AuthContext);
  if (!context) {
    throw new Error(`useAuth must be used within an AuthProvider`);
  }
  return context;
};
