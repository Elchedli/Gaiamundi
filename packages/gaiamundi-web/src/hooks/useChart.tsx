import { CHART_TYPES } from 'interfaces/chart';
import { useMemo } from 'react';
import { useQuery } from 'react-query';
import { getChartById } from 'services/chart';

export const useChart = (chartId: number) => {
  const query = useQuery({
    queryKey: ['chart', chartId],
    queryFn: async () => await getChartById(chartId),
    enabled: !!chartId,
  });
  const chart = query.data?.data;
  const ChartComponent = useMemo(() => {
    return chart?.type
      ? CHART_TYPES[chart.type].ChartComponent
      : CHART_TYPES.column.ChartComponent;
  }, [chart]);

  return {
    ...query,
    ChartComponent,
  };
};
