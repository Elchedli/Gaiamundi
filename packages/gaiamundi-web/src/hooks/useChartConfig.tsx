import { LoadingMessage } from 'components/Loader/LoadingMessage';
import { Dimensions, Direction } from 'eazychart-core';
import { ColumnChart } from 'eazychart-react';
import { ApiData } from 'interfaces/api';
import {
  AllChartConfigProps as AllChartProps,
  CHART_TYPES,
  Chart,
  RawDatum,
  RawDatumType,
} from 'interfaces/chart';
import React, {
  FC,
  ReactNode,
  createContext,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react';
import { useQuery, useQueryClient } from 'react-query';
import { getChartById } from 'services/chart';
import { getDataTypeMap, guessDomainKey } from 'utils/chart';
import { DEFAULT_DIMENSIONS, INITIAL_CHART_CONFIG } from 'utils/constants';

type ChartConfigContextType = {
  chart: ApiData<Chart>;
  ChartComponent: React.FC<AllChartProps>;
  updateChartProps: (_props: Partial<AllChartProps>) => void;
  updateChart: (config: Partial<Chart>) => void;
  rawData: RawDatum[];
  dataKeys: Record<string, RawDatumType>;
  dimensions: Dimensions;
  setDimensions: (_dimensions: Dimensions) => void;
};

const initialContext: ChartConfigContextType = {
  chart: INITIAL_CHART_CONFIG,
  ChartComponent: ColumnChart,
  updateChartProps: (_props: Partial<AllChartProps>) => undefined,
  updateChart: (_config: Partial<Chart>) => undefined,
  rawData: [],
  dataKeys: {},
  dimensions: DEFAULT_DIMENSIONS,
  setDimensions: (_dimensions: Dimensions) => undefined,
};

const ChartConfigContext =
  createContext<ChartConfigContextType>(initialContext);

type ChartConfigProviderProps = {
  chartId: number;
  rawData: RawDatum[];
  children: ReactNode;
};

export const ChartConfigProvider: FC<ChartConfigProviderProps> = ({
  chartId,
  rawData,
  children,
}) => {
  const [dimensions, setDimensions] = useState<Dimensions>(DEFAULT_DIMENSIONS);
  const queryClient = useQueryClient();
  const {
    data: response,
    error,
    isLoading,
  } = useQuery({
    queryKey: ['chart', chartId],
    queryFn: async () => await getChartById(chartId),
    enabled: !!chartId,
  });

  const chart = response?.data || INITIAL_CHART_CONFIG;
  const setChart = (newChart: ApiData<Chart> | undefined) => {
    queryClient.setQueryData(['chart', chartId], { data: newChart });
  };

  const updateChartProps = (props: Partial<AllChartProps>) => {
    setChart({
      ...chart,
      props: {
        ...chart.props,
        ...props,
      },
    });
  };

  const updateChart = (updates: Partial<Chart>) => {
    setChart({ ...chart, ...updates });
  };

  const ChartComponent = useMemo(() => {
    return chart?.type
      ? CHART_TYPES[chart.type].ChartComponent
      : CHART_TYPES.column.ChartComponent;
  }, [chart]);

  const dataKeys = useMemo(() => {
    if (rawData.length > 0) {
      return getDataTypeMap(rawData[0]);
    }
    return {};
  }, [rawData]);

  useEffect(() => {
    switch (chart.type) {
      case 'column':
        updateChartProps({
          xAxis: {
            ...chart.props.xAxis,
            domainKey: guessDomainKey(rawData, 'string', 0),
          },
          yAxis: {
            ...chart.props.yAxis,
            domainKey: guessDomainKey(rawData, 'number', 1),
          },
        });
        break;
      case 'lineColumn':
        updateChartProps({
          xAxis: {
            ...chart.props.xAxis,
            domainKey: guessDomainKey(rawData, 'string', 0),
          },
          yAxis: {
            ...chart.props.yAxis,
            domainKey: guessDomainKey(rawData, 'number', 1),
          },
        });
        break;
      case 'bar':
        updateChartProps({
          xAxis: {
            ...chart.props.xAxis,
            domainKey: guessDomainKey(rawData, 'number', 0),
          },
          yAxis: {
            ...chart.props.yAxis,
            domainKey: guessDomainKey(rawData, 'string', 1),
          },
        });
        break;
      case 'pie':
        updateChartProps({
          domainKey: guessDomainKey(rawData, 'number', 0),
        });
        break;
      case 'line':
        updateChartProps({
          xAxis: {
            ...chart.props.xAxis,
            domainKey: guessDomainKey(rawData, 'number', 0),
          },
          yAxis: {
            ...chart.props.yAxis,
            domainKey: guessDomainKey(rawData, 'number', 1),
          },
        });
        break;
      case 'area':
        updateChartProps({
          xAxis: {
            ...chart.props.xAxis,
            domainKey: guessDomainKey(rawData, 'number', 0),
          },
          yAxis: {
            ...chart.props.yAxis,
            domainKey: guessDomainKey(rawData, 'number', 1),
          },
        });
        break;
      case 'scatter':
        updateChartProps({
          xAxis: {
            ...chart.props.xAxis,
            domainKey: guessDomainKey(rawData, 'number', 0),
          },
          yAxis: {
            ...chart.props.yAxis,
            domainKey: guessDomainKey(rawData, 'number', 1),
          },
        });
        break;
      case 'bubble':
        updateChartProps({
          xAxis: {
            ...chart.props.xAxis,
            domainKey: guessDomainKey(rawData, 'number', 0),
          },
          yAxis: {
            ...chart.props.yAxis,
            domainKey: guessDomainKey(rawData, 'number', 1),
          },
        });
        break;

      default:
        break;
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [chart.type, dataKeys]);

  if (isLoading) {
    return <LoadingMessage />;
  }

  if (error) {
    return <>{error}</>;
  }

  return (
    <ChartConfigContext.Provider
      value={{
        chart,
        updateChartProps,
        updateChart,
        ChartComponent,
        dataKeys,
        rawData,
        dimensions,
        setDimensions,
      }}
    >
      {children}
    </ChartConfigContext.Provider>
  );
};

export const useChart = () => {
  return useContext(ChartConfigContext);
};

export const useChartGridConfig = () => {
  const { chart, updateChartProps } = useChart();

  const updateGridDirection = (direction: Direction, isEnabled: boolean) => {
    let gridDirections = [...chart.props.grid.directions] as Direction[];
    if (isEnabled) {
      gridDirections.push(direction);
    } else {
      gridDirections = gridDirections.filter((dir) => dir !== direction);
    }
    updateChartProps({
      grid: {
        ...chart.props.grid,
        directions: gridDirections,
      },
    });
  };

  const updateGridColor = (color: string) => {
    updateChartProps({
      grid: {
        ...chart.props.grid,
        color,
      },
    });
  };

  return {
    grid: chart.props.grid,
    updateGridDirection,
    updateGridColor,
  };
};
