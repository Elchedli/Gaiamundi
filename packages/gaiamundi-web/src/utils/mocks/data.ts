import { ApiCollection, ApiData } from 'interfaces/api';
import { DataFragment } from 'interfaces/data-fragment';
import { Dataset } from 'interfaces/dataset';
import { UploadedFile } from 'interfaces/file';
import { GeoMap } from 'interfaces/geo-map';
import { Indicator } from 'interfaces/indicator';
import { PageCarto } from 'interfaces/page-carto';
import { Tag } from 'interfaces/tag';
import { User } from 'interfaces/user';
import mockMapJson from './map.json';

export const mockMapPath = mockMapJson;

export const mockFileUpload = {
  id: 184,
  name: 'exampleParsing2.csv',
  alternativeText: null,
  caption: null,
  width: null,
  height: null,
  formats: null,
  hash: 'example_Parsing2_ce8756a044',
  ext: '.csv',
  mime: 'text/csv',
  size: 0.05,
  url: '/uploads/example_Parsing2_ce8756a044.csv',
  previewUrl: null,
  provider: 'local',
  provider_metadata: null,
  createdAt: '2023-03-16T18:20:09.476Z',
  updatedAt: '2023-03-16T18:20:09.476Z',
};

export const mockUser: User = {
  id: 1,
  username: 'testuser',
  password: 'test',
  email: 'test@test.com',
  provider: 'local',
  confirmed: true,
  blocked: false,
  created_at: '2022-02-07T00:00:00.000Z',
  updated_at: '2022-02-07T00:00:00.000Z',
};

export const mockGeoJson: ApiData<UploadedFile> = {
  id: 1,
  name: 'test.geojson',
  hash: 'abc123',
  provider: 'provider123',
  ext: '.geojson',
  mime: 'application/json',
  size: 1234,
  url: '/api/v1/geo-maps/id/thumbnail',
  formats: {
    thumbnail: {
      extension: 'ext',
      url: 'happyUrl',
      size: 360,
      hash: 'hashbar',
      mime: 'mimeExample',
    },
  },
  created_at: '2022-02-07T00:00:00.000Z',
  updated_at: '2022-02-07T00:00:00.000Z',
};

export const mockCover: ApiData<UploadedFile> = {
  id: 3,
  name: 'image.png',
  hash: 'hashImage',
  provider: 'provider123',
  ext: '.png',
  mime: 'application/image',
  size: 1234,
  url: '/api/v1/geo-maps/id/thumbnailImage',
  formats: {
    small: {
      extension: 'ext',
      url: '/api/v1/geo-maps/id/thumbnailImage?720',
      size: 360,
      hash: 'hash720',
      mime: 'mimeExample',
    },
    thumbnail: {
      extension: 'ext',
      url: '/api/v1/geo-maps/id/thumbnailImage?720',
      size: 360,
      hash: 'hash720',
      mime: 'mimeExample',
    },
  },
  created_at: '2022-02-07T00:00:00.000Z',
  updated_at: '2022-02-07T00:00:00.000Z',
};

export const mockGeoMap: ApiData<GeoMap> = {
  id: 1,
  name: 'Test Map',
  yearValidity: 2022,
  mesh: 'Test Maille',
  source: 'Test Source',
  license: 'Test License',
  properties: [
    {
      name: 'code',
      sample: '39',
      isGeoCode: true,
    },
    {
      name: 'nom',
      sample: 'Jura',
      isGeoCode: false,
    },
  ],
  owner: mockUser,
  geoJSON: mockGeoJson,
  created_at: '2022-02-07T00:00:00.000Z',
  updated_at: '2022-02-07T00:00:00.000Z',
};

export const mockTags: ApiData<Tag>[] = [
  {
    id: 1,
    name: 'Tag A',
    count: 5,
    created_at: '2022-02-07T00:00:00.000Z',
    updated_at: '2022-02-07T00:00:00.000Z',
  },
  {
    id: 2,
    name: 'Tag B',
    count: 3,
    created_at: '2022-02-07T00:00:00.000Z',
    updated_at: '2022-02-07T00:00:00.000Z',
  },
  {
    id: 3,
    name: 'Tag C',
    count: 2,
    created_at: '2022-02-07T00:00:00.000Z',
    updated_at: '2022-02-07T00:00:00.000Z',
  },
];

export const mockDataset: ApiData<Dataset> = {
  id: 122,
  name: 'firstDataset',
  origin: 3,
  isPublic: true,
  owner: mockUser,
  csv: {
    id: 5,
    name: 'file.csv',
    hash: 'hashCSV',
    provider: 'provider123',
    ext: '.csv',
    mime: 'application/csv',
    size: 1234,
    url: '/api/v1/Uploads/id/csvFile.csv',
    formats: {
      thumbnail: {
        extension: 'ext',
        url: '/api/v1/Uploads/id/csvFile.csv?720',
        size: 360,
        hash: 'hash720',
        mime: 'mimeExample',
      },
    },
    created_at: '2022-02-07T00:00:00.000Z',
    updated_at: '2022-02-07T00:00:00.000Z',
  },
  created_at: '2022-02-07T00:00:00.000Z',
  updated_at: '2022-02-07T00:00:00.000Z',
};

export const mockDataFragments: ApiData<DataFragment>[] = [
  {
    id: 1,
    name: 'FirstFragment',
    columns: [
      {
        name: 'firstColumn',
        sample: 5,
        source: 'source1',
        validity: '2021',
        isGeoCode: true,
      },
      {
        name: 'secondColumn',
        sample: 10.3,
        source: 'source2',
        validity: '2022',
        isGeoCode: false,
      },
      {
        name: 'thirdColumn',
        sample: 0.25,
        source: 'source3',
        validity: '2023',
        isGeoCode: false,
      },
      {
        name: 'fourthColumn',
        sample: 1000,
        source: 'source4',
        validity: '2024',
        isGeoCode: false,
      },
    ],
    dataset: mockDataset,
    created_at: '2022-02-07T00:00:00.000Z',
    updated_at: '2022-02-07T00:00:00.000Z',
  },
];

export const indicatorsMockData: ApiData<Indicator>[] = [
  {
    id: 4,
    name: 'mockIndicator1',
    description: 'This is a mock indicator',
    source: 'mock',
    validity: 2,
    created_at: '2023-03-14T10:06:41.979Z',
    updated_at: '2023-03-14T10:06:41.979Z',
    equation: 'A+2',
    variables: [],
  },
  {
    id: 5,
    name: 'mockIndicator2',
    description: 'This is a mock indicator',
    source: 'mock',
    validity: 4,
    variables: [],
    created_at: '2023-03-14T10:12:37.285Z',
    updated_at: '2023-03-14T10:12:37.285Z',
    equation: 'A*2',
  },
];

export const mockPageCartoData: ApiData<PageCarto> = {
  id: 1,
  name: 'Page Carto 1',
  owner: mockUser,
  tags: mockTags,
  map: mockGeoMap,
  cover: mockCover,
  indicators: indicatorsMockData,
  charts: [],
  html: '<p>Test HTML</p>',
  data_fragments: mockDataFragments,
  created_at: '2022-02-07T00:00:00.000Z',
  updated_at: '2022-02-07T00:00:00.000Z',
};

export const mockGeoMapData: ApiData<GeoMap> = {
  id: 1,
  name: 'Test Map',
  yearValidity: 2022,
  mesh: 'Test Maille',
  source: 'Test Source',
  license: 'Test License',
  properties: [
    {
      name: 'nom',
      sample: 'Taponas',
      isGeoCode: true,
    },
  ],
  owner: mockUser,
  geoJSON: {
    id: 1,
    name: 'test.geojson',
    hash: 'abc123',
    provider: 'provider123',
    ext: '.geojson',
    mime: 'application/json',
    size: 1234,
    url: 'https://example.com/api/v1/geo-maps/id/thumbnail',
    formats: {
      '720': {
        extension: 'ext',
        url: 'happyUrl',
        size: 360,
        hash: 'hashbar',
        mime: 'mimeExample',
      },
    },
    created_at: '2022-02-07T00:00:00.000Z',
    updated_at: '2022-02-07T00:00:00.000Z',
  },
  created_at: '2022-02-07T00:00:00.000Z',
  updated_at: '2022-02-07T00:00:00.000Z',
};

export const mockGeoMapApiCollection: ApiCollection<GeoMap> = {
  data: Array.from({ length: 3 }, (_, index) => ({
    ...mockGeoMapData,
    id: index + 1,
    name: `Test Map ${index + 1}`,
  })),
  meta: {
    pagination: {
      pageCount: 1,
      total: 3,
    },
  },
};
