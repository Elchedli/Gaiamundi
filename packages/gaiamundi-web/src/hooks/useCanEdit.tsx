import { useLocation } from 'react-router-dom';
import { useAuth } from './useAuth';

export const useCanEdit = () => {
  const { isAuthenticated } = useAuth();
  const location = useLocation();
  const isOnEditorPage = location.pathname.endsWith('edit');
  return isAuthenticated && isOnEditorPage;
};
