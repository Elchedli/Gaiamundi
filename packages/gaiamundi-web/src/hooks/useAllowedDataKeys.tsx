import { useMemo } from 'react';
import { ChartType, RawDatumType } from '../interfaces/chart';
import { useChartConfig } from './useChartConfig';
import { useData } from './useData';

export const ALLOWED_DOMAIN_TYPES_BY_CHART: {
  [key in ChartType]: Array<RawDatumType>;
} = {
  column: ['string', 'number'],
  bar: ['string', 'number'],
  scatter: ['number'],
  line: ['number'],
  lineColumn: ['string', 'number'],
  pie: ['number'],
  area: ['number'],
  bubble: ['number'],
};

export const useAllowedDataKeys = () => {
  const { chart: chartConfig } = useChartConfig();
  const { dataKeys } = useData();
  const allowedTypes = ALLOWED_DOMAIN_TYPES_BY_CHART[chartConfig.type];

  const domainKeyOptions = useMemo(
    () =>
      Object.entries(dataKeys)
        .filter(([, type]) => {
          return allowedTypes.indexOf(type) !== -1;
        })
        .map(([key]) => ({
          value: key,
          label: key,
        })),
    [allowedTypes, dataKeys]
  );

  return { domainKeyOptions };
};
