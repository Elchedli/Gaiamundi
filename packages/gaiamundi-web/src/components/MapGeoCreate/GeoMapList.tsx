import { useState } from 'react';
import classNames from 'classnames';
import { useQuery } from 'react-query';
import { ContentType, QueryParams, strapi } from 'services/strapi';
import GeoListItem from './GeoListItem';
import { Pagination } from 'components/Pagination/Pagination';
import { GeoMap } from 'interfaces/geo-map';
import { useAuth } from 'hooks/useAuth';

export const GeoMapList = () => {
  const paginationLimit = 9;

  const [page, setPage] = useState(1);
  const [selectedTab, setSelectedTab] = useState(0);
  const { user } = useAuth();

  const { data: response } = useQuery({
    queryKey: ['latest-geo-carto', page, selectedTab],
    queryFn: () => {
      return strapi.get<GeoMap>(ContentType.GEO_MAPS, {
        filters: {
          owner:
            !selectedTab && user != undefined
              ? {
                  id: {
                    $eq: user?.id,
                  },
                }
              : {},
        } as QueryParams['filters'],
        populate: '*',
        sort: 'createdAt:desc',
        pagination: {
          page: page,
          pageSize: paginationLimit,
        },
      });
    },
  });

  const GeoMap = [
    {
      id: 0,
      title: 'Mes cartes',
    },
    {
      id: 1,
      title: 'Toutes les cartes',
    },
  ];
  return (
    <div>
      <h2 className="pb-4 text-3xl font-extrabold">
        RÃ©utiliser une carte GeoJSON
      </h2>
      <div className="flex-column">
        <div className="flex bg-blue-600 rounded-t-lg p-2">
          {GeoMap.map((table, index) => (
            <div
              key={table.id}
              className={classNames(
                'rounded-t-lg p-4 text-sm font-medium leading-5 text-blue-700 cursor-pointer',
                'ring-blue ring-opacity-60 ring-offset-1 ring-offset-blue-400 focus:outline-none focus:ring-1',
                index === selectedTab
                  ? 'bg-white shadow'
                  : 'text-blue-100 hover:bg-white/[0.12] hover:text-white'
              )}
              onClick={() => setSelectedTab(index)}
            >
              {table.title}
            </div>
          ))}
        </div>
        <div className="rounded-b-lg border border-blue-700 p-5">
          <h2 className="pb-4 text-3xl font-extrabold">Toutes les cartes</h2>
          <div className="grid grid-cols-3 gap-y-10 gap-x-6">
            {response?.data.map((page) => {
              return <GeoListItem key={page.attributes.name} {...page} />;
            })}
          </div>
          <div className="flex flex-row mt-5 justify-center">
            <Pagination
              page={page}
              onPaginateNext={() => setPage(page + 1)}
              onPaginatePrevious={() => setPage(page - 1)}
              onPaginate={(p: number) => setPage(p)}
              totalPages={response?.meta.pagination.pageCount ?? 0}
            />
          </div>
        </div>
      </div>
    </div>
  );
};
