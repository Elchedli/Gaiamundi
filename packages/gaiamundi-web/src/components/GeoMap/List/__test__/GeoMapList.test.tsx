import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useAuth } from 'hooks/useAuth';
import { useQuery } from 'react-query';
import { mockGeoMapApiCollection } from 'utils/mocks/data';
import { GeoMapList } from '../GeoMapList';

jest.mock('hooks/useAuth');
jest.mock('react-query', () => ({
  useQuery: jest.fn(),
}));

describe('GeoMapList', () => {
  beforeEach(() => {
    // Mock useAuth hook to return a fake user object
    (useAuth as jest.Mock).mockReturnValue({
      isAuthenticated: true,
      user: { id: 1 },
    });
    // Mock the useQuery hook to return a fake response object
    (useQuery as jest.Mock).mockReturnValue({
      data: { ...mockGeoMapApiCollection },
    });
  });

  it('renders list of GeoListItem components and snapshot', async () => {
    const { container } = render(<GeoMapList />);
    const geoListItemElement = screen.getByText('Test Map 1');
    expect(geoListItemElement).toBeInTheDocument();
    expect(screen.getByText('Mes cartes')).toBeInTheDocument();
    expect(screen.getByText('Toutes les cartes')).toBeInTheDocument();
    expect(container.querySelectorAll('.grid > div').length).toBe(3);
    expect(container).toMatchSnapshot();
  });

  it('switches between tabs', async () => {
    const { container } = render(<GeoMapList />);
    const allMapsTab = screen.getByText('Toutes les cartes');
    userEvent.click(allMapsTab);
    expect(allMapsTab).toHaveClass('bg-white shadow');
    expect(container.querySelectorAll('.grid > div').length).toBe(3);
  });
});
