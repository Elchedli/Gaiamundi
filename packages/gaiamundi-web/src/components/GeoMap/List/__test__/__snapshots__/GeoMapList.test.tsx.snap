// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeoMapList renders list of GeoListItem components and snapshot 1`] = `
<div>
  <div
    class="flex-column"
  >
    <div
      class="flex bg-blue-600 rounded-t-lg p-2"
      data-testid="map-choice"
    >
      <div
        class="rounded-t-lg p-4 text-sm font-medium leading-5 text-blue-700 cursor-pointer ring-blue ring-opacity-60 ring-offset-1 ring-offset-blue-400 focus:outline-none focus:ring-1 bg-white shadow"
      >
        Mes cartes
      </div>
      <div
        class="rounded-t-lg p-4 text-sm font-medium leading-5 text-blue-700 cursor-pointer ring-blue ring-opacity-60 ring-offset-1 ring-offset-blue-400 focus:outline-none focus:ring-1 text-blue-100 hover:bg-white/[0.12] hover:text-white"
      >
        Toutes les cartes
      </div>
    </div>
    <div
      class="rounded-b-lg border border-blue-700 p-5"
    >
      <div
        class="grid grid-cols-3 gap-y-10 gap-x-6"
      >
        <div
          class="flex-col flex rounded-lg border border-gray-200 bg-white shadow-md dark:border-gray-700 dark:bg-gray-800 hover:bg-gray-100 dark:hover:bg-gray-700"
          data-testid="flowbite-card"
          id="mapId"
          role="card"
        >
          <div
            class="flex h-full flex-col justify-center gap-4"
          >
            <div
              class="group pb-2 cursor-pointer"
            >
              <div
                class="bg-slate-300"
              >
                <img
                  class="w-full object-contain object-center hover:opacity-75"
                  src="http://localhost:1337/api/geo-maps/thumbnail/1"
                />
              </div>
              <div
                class="p-3"
              >
                <h2>
                  Test Map 1
                </h2>
                <div
                  class="text-gray-500"
                >
                  <div
                    class="text-sm my-1"
                  >
                    <div
                      class="my-1 font-bold"
                    >
                      Source : 
                      Test Source
                    </div>
                    <div
                      class="text-xs mt-2"
                    >
                      License : 
                      Test License
                    </div>
                    <div
                      class="text-xs mt-2"
                    >
                      Annee de validitée : 
                      2022
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="flex-col flex rounded-lg border border-gray-200 bg-white shadow-md dark:border-gray-700 dark:bg-gray-800 hover:bg-gray-100 dark:hover:bg-gray-700"
          data-testid="flowbite-card"
          id="mapId"
          role="card"
        >
          <div
            class="flex h-full flex-col justify-center gap-4"
          >
            <div
              class="group pb-2 cursor-pointer"
            >
              <div
                class="bg-slate-300"
              >
                <img
                  class="w-full object-contain object-center hover:opacity-75"
                  src="http://localhost:1337/api/geo-maps/thumbnail/1"
                />
              </div>
              <div
                class="p-3"
              >
                <h2>
                  Test Map 2
                </h2>
                <div
                  class="text-gray-500"
                >
                  <div
                    class="text-sm my-1"
                  >
                    <div
                      class="my-1 font-bold"
                    >
                      Source : 
                      Test Source
                    </div>
                    <div
                      class="text-xs mt-2"
                    >
                      License : 
                      Test License
                    </div>
                    <div
                      class="text-xs mt-2"
                    >
                      Annee de validitée : 
                      2022
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="flex-col flex rounded-lg border border-gray-200 bg-white shadow-md dark:border-gray-700 dark:bg-gray-800 hover:bg-gray-100 dark:hover:bg-gray-700"
          data-testid="flowbite-card"
          id="mapId"
          role="card"
        >
          <div
            class="flex h-full flex-col justify-center gap-4"
          >
            <div
              class="group pb-2 cursor-pointer"
            >
              <div
                class="bg-slate-300"
              >
                <img
                  class="w-full object-contain object-center hover:opacity-75"
                  src="http://localhost:1337/api/geo-maps/thumbnail/1"
                />
              </div>
              <div
                class="p-3"
              >
                <h2>
                  Test Map 3
                </h2>
                <div
                  class="text-gray-500"
                >
                  <div
                    class="text-sm my-1"
                  >
                    <div
                      class="my-1 font-bold"
                    >
                      Source : 
                      Test Source
                    </div>
                    <div
                      class="text-xs mt-2"
                    >
                      License : 
                      Test License
                    </div>
                    <div
                      class="text-xs mt-2"
                    >
                      Annee de validitée : 
                      2022
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="flex flex-row mt-5 justify-center"
      >
        <div
          class="flex flex-row mt-5"
          data-testid="pagination"
        >
          <button
            class="group inline-flex h-min w-fit items-center justify-center p-0.5 text-center font-medium focus:z-10 rounded-lg text-white bg-blue-700 border border-transparent hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 disabled:hover:bg-blue-700 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800 dark:disabled:hover:bg-blue-600 cursor-not-allowed opacity-50 bg-gray-200"
            data-testid="button-element"
            disabled=""
            type="button"
          >
            <span
              class="flex items-center text-sm px-4 py-2 rounded-md"
            >
              1
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
